### DISCUSSION TITLE
RAG with LangchainJS: Session Resources

### DISCUSSION THREAD

**Discussion by multispark** [microsoft/RAG_Hack#40] [Announcements]
This is the resources and discussion thread for this live stream: https://aka.ms/raghack/js
It will be updated with links after the session.

Recording: https://www.youtube.com/watch?v=JCKXmABE7_U
Slides: https://github.com/microsoft/RAG_Hack/blob/main/slides/RAGHack-LangchainJS-Serverless.pptx
Blog: https://techcommunity.microsoft.com/t5/microsoft-developer-community/hack-together-rag-hack-building-rag-applications-with-langchain/ba-p/4238200

Please ask any follow-up questions here, and the speakers will see them.

**Comment by bhanukacodes** [comment]
Hi Team, thanks for the session.! I'm working on a chatbot project that implements the RAG pattern, and I've been referring to both the  javascript ([serverless-chat-langchainjs](https://github.com/Azure-Samples/serverless-chat-langchainjs), [azure-search-openai-javascript](https://github.com/Azure-Samples/azure-search-openai-javascript)) and python ([openai-demo](https://github.com/Azure-Samples/azure-search-openai-demo)) RAG blueprint projects. However, I have a few doubts and would greatly appreciate your feedback, specifically from the perspective of ensuring the best practices and recommendations for a production-grade application.

- For data ingestion, can we leverage integrated-vectorization by calling an AI Search API to index documents without using a custom implementation for handling document splitting, chunking, and indexing etc?
- Why is an orchestrator like LangChain framework is recommended compared to a simpler RAG implementation with Bot Framework?
- How can LangChain.js be used to preserve the state and context of a conversation to facilitate a context-aware conversational chat behavior?

**Reply by pamelafox** [reply]
I'll just reply to #1 since I have Python experience with the integrated vectorization feature, and let Yohan/Glaucia reply to the LangchainJS questions.
We do optionally provide a path in the azure-search-openai-demo repo for developers to use integrated vectorization. The benefits of that vectorization is that it's all cloud-based, can connect to many Azure stores, and can run on a schedule.

The drawbacks:
* The built-in splitting skill is based on characters, not token count. That's fine for English documents but not great if your documents are in CJK languages with much higher token:word ratios.
* The built-in indexing skill does not remember the PDF page number when it stores the page. That leads to a less optimal experience for the user since the citation can't jump to the relevant part of the page.
* Generally, it may be harder to customize if you run into other limitations.

**Reply by bhanukacodes** [reply]
Thanks for the detailed answer @pamelafox. 

Regarding your second point under drawbacks, I would assume that we could potentially add a custom skill to address that? However, as you mentioned, a custom ingestion implementation offers more flexibility.

**Reply by sinedied** [reply]
An orchestrator like LangChain is mainly useful for its ability to abstract away the difference between different AI models, Vector DBs and such. That means you can easily switch from one provider to another to make experiments or reduce costs, like we're doing in this session to switch between local models and DB and Azure models and DB. It also provides many components to load data from different sources if needed (YT videos, Office docs, GH repos, etc).

For preserving context, Langchain.js has this concept of ["memory"](https://js.langchain.com/v0.2/docs/integrations/memory) that you can use to store/restore chat history and use it to answer questions. You can see an example here: https://js.langchain.com/v0.2/docs/how_to/qa_chat_history_how_to/#chain-with-chat-history

It's not implemented yet in our JS sample, but it's in our backlog.

**Reply by pamelafox** [reply]
@bhanukacodes Yep, you should be able to bring in a custom skill instead, if you're comfortable building skills. I believe they're implemented as Azure Functions.

**Reply by bhanukacodes** [reply]
Thanks you very much @sinedied and @pamelafox for the valuable information shared.!

**Comment by pamelafox** [comment]
Here's a great blog post from Glaucia summarizing the session:
https://techcommunity.microsoft.com/t5/microsoft-developer-community/hack-together-rag-hack-building-rag-applications-with-langchain/ba-p/4238200

**Comment by DMK1961** [comment]
In Github Codespaces, After entering 'azd up' I am getting the following error:
---
ERROR: error executing step command 'provision': deployment failed: error deploying infrastructure: deploying to subscription:

Deployment Error Details:
InvalidTemplateDeployment: The template deployment 'openai' is not valid according to the validation procedure. The tracking id is 'a7e493c3-b851-4a25-b827-4df3708a6b77'. See inner errors for details.
InsufficientQuota: This operation require 15 new capacity in quota Tokens Per Minute (thousands) - GPT-4-Turbo, which is bigger than the current available capacity 8. The current quota usage is 0 and the quota limit is 8 for quota Tokens Per Minute (thousands) - GPT-4-Turbo.

TraceID: e450e5023599b0d113ad04506ee663e9
---

I have submitted a quota increase request. Not sure whether the same will be processed in time. 

Thanks.

**Reply by sinedied** [reply]
In the meantime, you can try reducing the capacity to `8` here: https://github.com/Azure-Samples/serverless-chat-langchainjs/blob/main/infra/main.bicep#L45 and redeploying, this should fix your issues.

Once your quota has been updated, you can safely change the capacity and redeploy again.

**Reply by DMK1961** [reply]
Thanks. Yes.

**Comment by DMK1961** [comment]
Yohan,
Thanks for your prompt reply.
Will check it out at the earliest.
DMK1961
    On Monday 16 September, 2024 at 05:37:20 pm IST, Yohan Lasorsa ***@***.***> wrote:  
 
 


In the meantime, you can try reducing the capacity to 8 here: https://github.com/Azure-Samples/serverless-chat-langchainjs/blob/main/infra/main.bicep#L45 and redeploying, this should fix your issues.

Once your quota has been updated, you can safely change the capacity and redeploy again.

**Reply by DMK1961** [reply]
Yohan,

Thanks a lot. That worked.

**Comment by negar-mir** [comment]
Hi @sinedied 
I wanted to follow the tutorial for my learning. in [creating-a-new-azure-functions-project  ](https://github.com/Azure-Samples/serverless-chat-langchainjs/blob/main/docs/tutorial/02-setting-up-azure-functions.md#creating-a-new-azure-functions-project) suggested to start with starter branch but I can't see it in the list of branches. 
![image](https://github.com/user-attachments/assets/098d91c7-36cf-4c7f-8bbb-f38f89e28153)
could you please check how to access starter branch?

Best regards,
Negar

**Reply by sinedied** [reply]
Hi! The tutorial is currently work in progress and not up to date, that's why the `starter` branch does not exist yet.
@glaucia86 is currently working on it.

Meanwhile, it's not as detailed, but we have this [blog post](https://dev.to/azure/build-a-serverless-chatgpt-with-rag-using-langchainjs-3487) explaining how the sample works.

**Reply by negar-mir** [reply]
thanks
